readthedocs/core/views/__init__.py

```diff
@@ -124,7 +129,36 @@ def server_error_404(request, exception=None, template_name='404.html'):  # pyli
             )
         else:
             return response
-    r = render(request, template_name)
+
+    project, path = project_and_path_from_request(request, request.get_full_path())
+    version_slug = project.get_default_version()
+    version = project.versions.get(slug=version_slug)
+
+    filename = resolve_path(
+        project,
+        version_slug=version_slug,
+        filename='404.html',
+        subdomain=True,  # subdomain will make it a "full" path without a URL prefix
+    )
+
+    # This breaks path joining, by ignoring the root when given an "absolute" path
+    if filename[0] == '/':
+        filename = filename[1:]
+
+    if version.privacy_level == PRIVATE:
+        symlink = PrivateSymlink(project)
+    else:
+        symlink = PublicSymlink(project)
+    basepath = symlink.project_root
+    fullpath = os.path.join(basepath, filename)
+
+    if os.path.exists(fullpath):
+        log.debug('Serving custom 404.html page for project: %', project.slug)
+        r = HttpResponse(open(fullpath).read(), content_type='text/html')
```

Comment:
I'm not sure if we want to serve this file from Django or we prefer to use the `X-Accel-Redirect` header here.

Reply:
In fact, I think this is a bug (or a potential bug). I did the same at https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/core/views/serve.py#L269 but, what happen if this file is too big, for example?

I suppose that we should rely on NGINX to serve them.

I'd like to get some opinions here.

Reply:
The main issue here is that it doesn't work locally. We should likely have it do Nginx when DEBUG=False, and serve locally when DEBUG=True.

Reply:
Actually, it should perhaps respect PYTHON_MEDIA instead of DEBUG

Reply:
Well, 404 handler is not called at all on `DEBUG=True`, so we will need `DEBUG=False` anyway to get at this point.

I agree that it should rely on `PYTHON_MEDIA` only.

Reply:
I'm using `readthedocs.core.views.serve._serve_file` that does this for us ;)